File: backend/dataset/maxPathSum/maxPathSum_17.py
import math
class Solution:
    def helper(self,root):
        if root==None:
            return(-math.inf,-math.inf)
        l,lmax=self.helper(root.left)
        r,rmax=self.helper(root.right)
        c=max(root.val,l+root.val,r+root.val)
        s=max(c,lmax,rmax,l+r+root.val)
        return(c,s)
    def maxPathSum(self, root: TreeNode) -> int:
        return self.helper(root)[1]

File: backend/CDataset/maxPathSum/maxPathSum_17.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int currentSum = INT_MIN;
        int maxSum = INT_MIN;
        
        go(root,currentSum,maxSum);
        
        return maxSum;
    }
    
    int go(TreeNode* root,int &currentSum,int &maxSum){
        
        int l = 0;
        int r = 0;
        
        if (root->left != nullptr){
            l = go(root->left,currentSum,maxSum);
        }
        if (root->right != nullptr){
            r = go(root->right,currentSum,maxSum);
        }
        
        currentSum = max(l + root->val,r + root->val);
        currentSum = max(currentSum,root->val);
        maxSum = max(max(maxSum,currentSum),l + r + root->val);
       
        
        return currentSum;
    }
};
